File: Algorithms/102_BinaryTreeLevelOrderTraversal.java
public class Solution {
    public ArrayList<ArrayList<Integer>> levelOrder(TreeNode root) {
        ArrayList<ArrayList<Integer>> result = new ArrayList<ArrayList<Integer>>();
        if (root == null) {
            return result;
        }
        ArrayList<Integer> v = new ArrayList<Integer>();
        ArrayList<TreeNode> stack = new ArrayList<TreeNode>();
        ArrayList<TreeNode> next_level_stack = new ArrayList<TreeNode>();
        stack.add(root);
        while (stack.size() != 0) {
            TreeNode p = stack.get(0);
            stack.remove(0);
            v.add(p.val);
            if (p.left != null) {
                next_level_stack.add(p.left);
            }
            if (p.right != null) {
                next_level_stack.add(p.right);
            }
            if (stack.size() == 0) {
                stack = next_level_stack;
                next_level_stack = new ArrayList<TreeNode>();
                result.add(v);
                v = new ArrayList<Integer>();
            }
        }
        return result;
    }
}

File: Algorithms/102.BinaryTreeLevelOrderTraversal.cpp
class Solution {
public:
    vector<vector<int> > levelOrder(TreeNode *root) {
        vector<vector<int> > result;
        if (root == NULL) {
            return result;
        }
        vector<int> v;
        vector<TreeNode*> stack;
        vector<TreeNode*> next_level_stack;
        stack.push_back(root);
        while (!stack.empty()) {
            TreeNode* p = stack.front();
            stack.erase(stack.begin());
            v.push_back(p->val);
            if (p->left != NULL) {
                next_level_stack.push_back(p->left);
            }
            if (p->right != NULL) {
                next_level_stack.push_back(p->right);
            }
            if (stack.empty()) {
                stack = next_level_stack;
                next_level_stack.clear();
                result.push_back(v);
                v.clear();
            }
        }
        return result;
    }
};
